hypothesis-python/src/hypothesis/internal/constants_ast.py

```diff
@@ -166,33 +178,46 @@ def visit_Constant(self, node):
         self.generic_visit(node)
 
 
-def _constants_from_source(source: Union[str, bytes]) -> Constants:
+def _constants_from_source(source: Union[str, bytes], *, limit: bool) -> Constants:
     tree = ast.parse(source)
-    visitor = ConstantVisitor()
-    visitor.visit(tree)
+    visitor = ConstantVisitor(limit=limit)
+
+    try:
+        visitor.visit(tree)
+    except TooManyConstants:
+        # in the case of an incomplete collection, return nothing, to avoid
+        # muddying caches etc.
+        return Constants()
+
     return visitor.constants
 
 
 @lru_cache(4096)
-def constants_from_module(module: ModuleType) -> Constants:
+def constants_from_module(module: ModuleType, *, limit: bool = True) -> Constants:
     try:
         module_file = inspect.getsourcefile(module)
         # use type: ignore because we know this might error
         source_bytes = Path(module_file).read_bytes()  # type: ignore
     except Exception:
         return Constants()
 
+    if limit and len(source_bytes) > 512 * 1024:
+        # Skip files over 512kb. For reference, the largest source file
+        # in Hypothesis is strategies/_internal/core.py at 107kb at time
+        # of writing.
+        return Constants()
+
     source_hash = hashlib.sha1(source_bytes).hexdigest()[:16]
     cache_p = storage_directory("constants") / source_hash
```

Comment:
```suggestion
    cache_p = storage_directory("constants") / source_hash + ("nolimit" * limit)  # sketch example
```
easiest format, just have a separate cache file for each case.  This avoids paying the parsing cost when we've written unlimited cache files, but only want a limited amount for fast startup.